client/components/tinymce/plugins/wpcom-view/wpvideo-view.jsx

```diff
@@ -0,0 +1,137 @@
+
+/**
+ * Internal dependencies
+ */
+import shortcodeUtils from 'lib/shortcode';
+import { getVideo } from 'state/videos/selectors';
+import QueryVideo from 'components/data/query-video';
+
+class WpVideoView extends Component {
+
+	static match( content ) {
+		const match = shortcodeUtils.next( 'wpvideo', content );
+
+		if ( match ) {
+			return {
+				index: match.index,
+				content: match.content,
+				options: {
+					shortcode: match.shortcode
+				}
+			};
+		}
+	}
+
+	static serialize( content ) {
+		return encodeURIComponent( content );
+	}
+
+	constrainVideoDimensions( shortcodeWidthAttribute, shortcodeHeightAttribute ) {
+		let { width, height } = this.props.video
+			? this.props.video
+			: {
+				width: 640,
+				height: 640 / ( 16 / 9 )
+			};
+		const aspectRatio = width / height;
+
+		if ( shortcodeWidthAttribute && ! shortcodeHeightAttribute ) {
+			width = shortcodeWidthAttribute;
+			height = shortcodeWidthAttribute / aspectRatio;
+		} else if ( ! shortcodeWidthAttribute && shortcodeHeightAttribute ) {
+			width = shortcodeHeightAttribute * aspectRatio;
+			height = shortcodeHeightAttribute;
+		} else if ( shortcodeWidthAttribute && shortcodeHeightAttribute ) {
+			width = shortcodeWidthAttribute;
+			height = shortcodeHeightAttribute;
+		}
+
+		return { width, height };
+	}
+
+	getVideoAttributes() {
+		const shortcode = shortcodeUtils.parse( this.props.content );
+		const namedAttrs = shortcode.attrs.named;
+
+		const videoDimensions = this.constrainVideoDimensions(
+			parseInt( namedAttrs.w, 10 ),
+			parseInt( namedAttrs.h, 10 )
+		);
+
+		const shortcodeAttributes = {
+			guid: shortcode.attrs.numeric[0],
+			w: videoDimensions.width,
+			h: videoDimensions.height,
+			autoplay: includes( [ 'true', '1' ], namedAttrs.autoplay ),
+			hd: includes( [ 'true', '1' ], namedAttrs.hd ),
+			loop: includes( [ 'true', '1' ], namedAttrs.loop ),
+			at: parseInt( namedAttrs.at, 10 ) || 0,
+			defaultLangCode: namedAttrs.defaultlangcode
+		};
+
+		return Object.assign(
+			pick( shortcodeAttributes, [ 'guid', 'w', 'h' ] ),
+			{ embedUrl: this.getEmbedUrl( shortcodeAttributes ) }
+		);
+	}
+
+	getEmbedUrl( shortcodeAttributes ) {
+		const defaultAttributeValues = { hd: false, at: 0, defaultLangCode: undefined };
+		const attributesWithNonDefaultValues = omitBy(
+			shortcodeAttributes,
+			( value, key ) => defaultAttributeValues[ key ] === value );
+		const queryStringAttributes = [ 'autoplay', 'hd', 'loop', 'at', 'defaultLangCode' ];
+		const queryString = QueryString.stringify( pick( attributesWithNonDefaultValues, queryStringAttributes ) );
+
+		return `https://videopress.com/embed/${ shortcodeAttributes.guid }?${ queryString }`;
+	}
+
+	onLoad() {
+		const doc = this.props.editor.iframeElement.contentDocument;
+		const script = doc.createElement( 'script' );
+		script.src = 'https://videopress.com/videopress-iframe.js';
+		script.type = 'text/javascript';
+		doc.getElementsByTagName( 'head' )[0].appendChild( script );
```

Comment:
>Another consequence of adding this event listener is that by removing and re-adding the script the registered event listener is left behind so those do pile up.

And each of those event handlers is executed when switching from HTML to Visual mode? I seem to recall this being the case, but that the script itself was smart enough not to duplicate work if it had already detected that it was previously initialized.

If it is worrysome (which it very well might not be), another avenue to pursue could be to render the view within its own `iframe`, such that there would only ever be one script per `iframe`. Manipulating markup of frames, especially with concern to page lifecycle events (`DOMContentLoaded`) is sometimes easier said than done, but it is do-able. See [`<ShortcodeFrame />`](https://github.com/Automattic/wp-calypso/blob/master/client/components/shortcode/frame.jsx) and [`document.write`](https://developer.mozilla.org/en-US/docs/Web/API/Document/write) for more information if you're curious.